### PR TITLE
add import from JSON for ALT Linux distributions

### PR DESCRIPTION
**Pull request by dshein-alt**
- update sisyphus.yaml
- add SisyphusJsonParser in sisyphus.py

### CODE REVIEW

File: repology/parsers/parsers/sisyphus.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                    f"{packagedata['epoch']}:{packagedata['version']}-{packagedata['release']}"
+                )
+                pkg.add_categories(packagedata['category'])
+                pkg.add_homepages(packagedata['url'])
+                pkg.set_summary(packagedata['summary'])
+                pkg.add_licenses(packagedata['license'])
+                pkg.add_maintainers(packagedata['packager'])
+                # store package epoch and release
+                pkg.set_extra_field('epoch', packagedata['epoch'])
+                pkg.set_extra_field('release', packagedata['release'])
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:24]: This is not needed. `extrafields` are only used as a way to pass some extra data to packaglink templates, and versions are not used there.

File: repology/parsers/parsers/sisyphus.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                # store source package binaries
+                # XXX: is there any way to store binary packages extra fields?
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:26]: Yes, by generating additional binary packages, but there's no much sense to add them when src packages are available. Filling in `binnames` is quite enough here.

File: repology/parsers/parsers/sisyphus.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                pkg.add_binnames(
+                    [b['name'] for b in packagedata['binaries']]
+                )
+                # set package links
+                pkg.add_links(LinkType.PACKAGE_HOMEPAGE, packagedata['homepage'])
+                pkg.add_links(LinkType.PACKAGE_RECIPE_RAW, packagedata['recipe'])
+                pkg.add_links(LinkType.PACKAGE_ISSUE_TRACKER, packagedata['bugzilla'])
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:33]: Links must be filled either here or in yaml's `packagelinks`, not both.

As soon as there's no link for `PACKAGE_RECIPE` in JSON (e.g. `https://packages.altlinux.org/en/sisyphus/srpms/<name>/specfiles/`) I'd prefer 4 links to be filled in `packagelinks`. If the link is added to the json, all 4 links may be filled here.

File: repology/parsers/parsers/sisyphus.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from typing import Iterable
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:1]: Lacks GPL header. Please make sure not to copypaste my name there.

File: repology/parsers/parsers/sisyphus.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+from repology.packagemaker import PackageFactory, PackageMaker, LinkType
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:3]: See actions run log for flake8 errors.

**Review by AMDmi3 — CHANGES_REQUESTED** [review]

File: repology/parsers/parsers/sisyphus.py (hunk 1 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: done

File: repology/parsers/parsers/sisyphus.py (hunk 2 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: done

File: repology/parsers/parsers/sisyphus.py (hunk 3 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py:46]: done

File: repology/parsers/parsers/sisyphus.py (hunk 4 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py:18]: done

File: repology/parsers/parsers/sisyphus.py (hunk 5 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: done

**Comment by dshein-alt** [comment]
all fixes in 2a3974e

**Comment by dshein-alt** [comment]
Fixed missing name_type argument in parser.

Not sure but I guess that GENERIC_GEN_NAME is suitable here.

**Comment by dshein-alt** [comment]
fixed import order

File: repology/parsers/parsers/sisyphus.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                pkg.add_binnames(
+                    [b['name'] for b in packagedata['binaries']]
+                )
+                # set package links
+                pkg.add_links(LinkType.PACKAGE_HOMEPAGE, packagedata['homepage'])
+                pkg.add_links(LinkType.PACKAGE_RECIPE_RAW, packagedata['recipe'])
+                pkg.add_links(LinkType.PACKAGE_ISSUE_TRACKER, packagedata['bugzilla'])
+                pkg.add_links(
+                    LinkType.PACKAGE_RECIPE,
+                    packagedata['recipe'].rsplit('/', maxsplit=1)
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:49]: This does not make sense. Either all URLs are provided in the dump and Repology uses them as is and does not need to know anything about their internals, or all URLs are constructed by Repology.

**Review by AMDmi3 — CHANGES_REQUESTED** [review]
Also please squash into a single commit.

File: repology/parsers/parsers/sisyphus.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+from repology.packagemaker import NameType, LinkType, PackageFactory, PackageMaker
+from repology.parsers import Parser
+from repology.parsers.json import iter_json_list
+
+
+class SisyphusJsonParser(Parser):
+    def iter_parse(self, path: str, factory: PackageFactory) -> Iterable[PackageMaker]:
+        for packagedata in iter_json_list(path, ('packages', None)):
+            with factory.begin() as pkg:
+                pkg.add_name(packagedata['name'], NameType.GENERIC_GEN_NAME)
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:29]: NameType.SRCRPM_NAME is suitable here.

**Review by AMDmi3 — CHANGES_REQUESTED** [review]

File: repology/parsers/parsers/sisyphus.py [hunk 8]
```diff
@@ -44,10 +44,6 @@ def iter_parse(self, path: str, factory: PackageFactory) -> Iterable[PackageMake
                 pkg.add_links(LinkType.PACKAGE_HOMEPAGE, packagedata['homepage'])
                 pkg.add_links(LinkType.PACKAGE_RECIPE_RAW, packagedata['recipe'])
                 pkg.add_links(LinkType.PACKAGE_ISSUE_TRACKER, packagedata['bugzilla'])
-                pkg.add_links(
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py]: But `PACKAGE_RECIPE` link is still required.

**Review by AMDmi3 — CHANGES_REQUESTED** [review]

File: repology/parsers/parsers/sisyphus.py (hunk 8 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: The case is that if we use full link it leads to download spec file directly.
But without file name at the end (as it was stripped in previous commit) it leads to display spec file on our site.

So should I just put  the spec file download link copy to  PACKAGE_RECIPE here?

File: repology/parsers/parsers/sisyphus.py (hunk 7 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: In fact "name" field in our JSON contains source package name excluding epoch, version, release, architecture etc.

File: repology/parsers/parsers/sisyphus.py (hunk 8 above)
Comment (AMDmi3) [reply repology/parsers/parsers/sisyphus.py]: PACKAGE_RECIPE is intended for humans and PACKAGE_RECIPE_RAW is intended for automated tools. These are not interchangeable and both are required. I suggest to either put the missing link to the dump, or switch to `packagelinks`. In fact, I'd prefer the latter as it gives more flexibility.

File: repology/parsers/parsers/sisyphus.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with repology.  If not, see <http://www.gnu.org/licenses/>.
+
+from typing import Iterable
+
+from repology.packagemaker import LinkType, NameType, PackageFactory, PackageMaker
```
Comment (AMDmi3) [inline repology/parsers/parsers/sisyphus.py:20]: LinkType comes from `repology.package`, that's the cause for build failure.

File: repology/parsers/parsers/sisyphus.py (hunk 8 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: I'll update our API export code to provide both links for PACKAGE_RECIPE and PACKAGE_RECIPE_RAW separately.

Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: JSON schema updated at API.
Parser ready.

File: repology/parsers/parsers/sisyphus.py (hunk 9 above)
Comment (dshein-alt) [reply repology/parsers/parsers/sisyphus.py]: done

**Review by AMDmi3 — APPROVED** [review]

**Comment by AMDmi3** [comment]
It's live, thank you for your work!

Number of -unclassified packages had decreased significantly because upstream URLs are now available. I still need to look through remaining ones.

BTW, it would be nice to add more URLs if possible, such as build logs and patches. I have plans on new features based on these.

**Comment by dshein-alt** [comment]
In case of build logs, patches etc. - we will try to add such info but I can't say when. Currently such information is contained in build tasks from Hasher (reproducible building environment) but not directly in DB that API is build on.

Thank you!

**Pull request merged**